### PR TITLE
Properly group multiple inherent `impl`s

### LINKED ISSUE #251: Group items even if multiple inherent `impl`s are used
```rust
pub struct pdbtbx::PDB
pub struct field pdbtbx::PDB::identifier: core::option::Option<alloc::string::String>
... (more fields)
impl pdbtbx::PDB
impl pdbtbx::PDB
impl pdbtbx::PDB
impl<'a> pdbtbx::PDB
```

This example is from the crate `pdbtbx` where I have used multiple impl blocks on a single struct to allow fro grouping of functions with some general comments about those groups. That is why it stands out so much. But I would assume these impl blocks (just the main block with all attached functions) to not be shown in the list here. An argument could be made to have the functions grouped below their respective impl blocks, especially when generic bounds are used. But this seems to not be the case here.

### PR DESCRIPTION
**Pull request by Enselic**
"properly" meaning "group items like rustdoc HTML does it", which means each inherent `impl` is presented separately (i.e. no de-duplication)

Closes #251

Depends on #269

### CODE REVIEW

**Comment by Enselic** [comment]
Had to rebase to trigger CI, but we should not get this problem in the future (see https://github.com/Enselic/cargo-public-api/pull/269#issuecomment-1374435468)

**Comment by Emilgardis** [comment]
Does moving a function from one impl to another trigger a change with `cargo public-api diff --deny changes` ?

**Comment by Enselic** [comment]
Good question! I wasn't sure at first, so I wrote a test for it, which I have now pushed. Turns out there is no diff for this case. IMHO that is correct since the public API has not changed.

The reason there is no diff triggered is because the diffing algo is based on HashBags. So the sorted order of items does not matter. (The first version of the algorithm depended on sort order however, but about half a year ago the diff algo was rewritten to not depend on sort order.)

**Comment by Enselic** [comment]
Wait, I forgot to make the `fn`s `pub`, and when I fix that there is a diff that shows up :scream: Need to investigate a bit more here...

**Comment by Enselic** [comment]
Since the "impl name" of the moved item changes, a diff is triggered. I think we need to somehow differentiate between a "sortable path" and a "ID path". Might require some significant conceptual changes...

**Comment by Enselic** [comment]
One easy way to solve this is to make `Eq` and `Hash` only take `PublicItem::tokens` into account. When diffing, an item that *looks* the same (and from an API point of view *is* the same) will then be considered equal. In that sense, I like that solution.

The downside is that now two items might be `Ordering::Equal` but not `eq` to each other. I'm unsure if this is a problem. It is clearly stated that `Eq` and `Hash` needs to match, but I see no such thing for `Ord`.

So maybe it is fine to do it like that? If yes, the PR in its current state is production ready.

But if `Ord` should/must be in sync with `Eq`, I think we should make `Ord` consistent with `Eq`/`Hash` and instead use a special key/method when sorting. I'm unsure what the best way forward is here.

**Comment by Emilgardis** [comment]
What do you mean with `Cmp`?

**Comment by Enselic** [comment]
Sorry, I meant `Ord` (which has a `cmp` method)

**Comment by Emilgardis** [comment]
If you mean `PartialOrd::partial_cmp` and by extension `Ord::cmp` then yes, it's a violation of

> 1. a == b if and only if partial_cmp(a, b) == Some(Equal).

**Comment by Enselic** [comment]
Ah right, [there](https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html) it is, thanks

I pushed a commit now that should make things well again

**Comment by Emilgardis** [comment]
yes, my bad I could've linked it!

File: public-api/src/intermediate_public_item.rs [hunk 1]
```diff
@@ -42,6 +42,46 @@ impl<'c> NameableItem<'c> {
             sortable_name.push_str(&crate::tokens::tokens_to_string(
                 &context.render_impl(impl_, &[]),
             ));
+
+            // If this is an inherent impl, additionally add the concatenated
+            // names of all associated items to "name" of the impl. This makes
```
Comment (Emilgardis) [inline public-api/src/intermediate_public_item.rs]: possible typo?, `to the "name"`

File: public-api/src/intermediate_public_item.rs [hunk 2]
```diff
@@ -42,6 +42,46 @@ impl<'c> NameableItem<'c> {
...
+            //
+            // In this case, we want to group the two impls together. So make
+            // the name of the first impl to be
+            //
+            //   impl MultipleInherentImpls-impl_one
+            //
+            // and the second one to be
+            //
+            //   impl MultipleInherentImpls-impl_two
+            //
```
Comment (Emilgardis) [inline public-api/src/intermediate_public_item.rs]:
```suggestion
            // In this case, we want to group the two impls together. So
            // the name of the first impl should be
            //
            //   impl MultipleInherentImpls-impl_one
            //
            // and the second one
            //
            //   impl MultipleInherentImpls-impl_two
            //
```

File: public-api/src/public_item.rs [hunk 3]
```diff
@@ -40,6 +41,34 @@ impl PublicItem {
     pub fn tokens(&self) -> impl Iterator<Item = &Token> {
         self.tokens.iter()
     }
+
+    /// Special version of `cmp` that is used to sort public items in a way that
```
Comment (Emilgardis) [inline public-api/src/public_item.rs]:
```suggestion
    /// Special version of [`cmp`](Ord::cmp) that is used to sort public items in a way that
```

**Review by Emilgardis — APPROVED** [review]
lgtm! some minor changes/suggestions

File: public-api/src/intermediate_public_item.rs (hunk 1 above)
Comment (Enselic) [reply public-api/src/intermediate_public_item.rs]: fixed!

**Comment by Emilgardis** [comment]
feel free to squash without adding me as co-author :) or merge/squash merge as is :D

**Comment by Enselic** [comment]
The commit history ended up too messy for my taste, so I force-push a better commit structure (where each commit passes CI). No code changes were made with the force-push. Soon merging, thanks :)

**Pull request merged**